src/Picasso_MarchingCubes.hpp

```diff
@@ -0,0 +1,650 @@
+    for ( int d = 0; d < 3; ++d )
+    {
+        ++index[d];
+    }
+    local_mesh.coordinates( Cajita::Node{}, index,
+                            vertex_locations.data() + 3 );
+    for ( int d = 0; d < 3; ++d )
+    {
+        vertex_locations[d + 3] -= vertex_locations[d];
+    }
+}
+
+//---------------------------------------------------------------------------//
+// Compute the edges.
+KOKKOS_INLINE_FUNCTION
+void computeEdges( const Kokkos::Array<double, 8>& vertex_data,
+                   const Kokkos::Array<int, 8>& vertex_signs,
+                   Kokkos::Array<Kokkos::Array<double, 3>, 12>& edges )
+{
+    // Calculate edge weights via linear interpolation.
+    auto compute_weight = [&]( const int v1, const int v2 ) {
+        return ( vertex_signs[v1] != vertex_signs[v2] )
+                   ? vertex_data[v1] / ( vertex_data[v1] - vertex_data[v2] )
+                   : -1.0;
+    };
+
+    // z edges
+    edges[0][0] = 0.0;
+    edges[0][1] = 0.0;
+    edges[0][2] = compute_weight( 0, 1 );
+
+    edges[1][0] = 0.0;
+    edges[1][1] = 1.0;
+    edges[1][2] = compute_weight( 2, 3 );
+
+    edges[2][0] = 1.0;
+    edges[2][1] = 0.0;
+    edges[2][2] = compute_weight( 4, 5 );
+
+    edges[3][0] = 1.0;
+    edges[3][1] = 1.0;
+    edges[3][2] = compute_weight( 6, 7 );
+
+    // y edges
+    edges[4][0] = 0.0;
+    edges[4][1] = compute_weight( 0, 2 );
+    edges[4][2] = 0.0;
+
+    edges[5][0] = 0.0;
+    edges[5][1] = compute_weight( 1, 3 );
+    edges[5][2] = 1.0;
+
+    edges[6][0] = 1.0;
+    edges[6][1] = compute_weight( 4, 6 );
+    edges[6][2] = 0.0;
+
+    edges[7][0] = 1.0;
+    edges[7][1] = compute_weight( 5, 7 );
+    edges[7][2] = 1.0;
+
+    // x edges
+    edges[8][0] = compute_weight( 0, 4 );
+    edges[8][1] = 0.0;
+    edges[8][2] = 0.0;
+
+    edges[9][0] = compute_weight( 1, 5 );
+    edges[9][1] = 0.0;
+    edges[9][2] = 1.0;
+
+    edges[10][0] = compute_weight( 2, 6 );
+    edges[10][1] = 1.0;
+    edges[10][2] = 0.0;
+
+    edges[11][0] = compute_weight( 3, 7 );
+    edges[11][1] = 1.0;
+    edges[11][2] = 1.0;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+  \brief Build the marching cubes triangulation over the owned cells. This
+  requires the signed distance values to have been gathered.
+  \param exec_space The execution space to use for parallel operations.
+  \param signed_distance The signed distance array from which to build the
+  facets. This array must be on the nodes as this is a cell-based operation.
+  \param data The resulting facet data.
+  \param reset_data_memory If true, the facet data will always be
+  resized/reallocated. If false, reallocation will only occur when the facet
+  or cell count has exceeded the existing count limits.
+*/
+template <class ExecutionSpace, class Mesh, class SignedDistanceArray,
+          class MemorySpace>
+void build( const ExecutionSpace& exec_space, const Mesh& mesh,
```

Comment:
I haven't done this in any of our codes so far so I don't think I want to do it here. A user could pass in a memory space instance here and it would just use the default execution space for that memory space.